File: app/src/main/java/omniex/nl/omniex/ui/app/product/list/ProductsListPresenter.java
package omniex.nl.omniex.ui.app.product.list;

import javax.inject.Inject;

import nl.omniex.omniexshopping.data.repository.ProductsRepository;
import nl.omniex.omniexshopping.ui.base.BasePresenter;
import nl.omniex.omniexshopping.ui.base.BaseView;
import timber.log.Timber;

public class ProductsListPresenter extends BasePresenter<ProductsListView> {

    private ProductsRepository mProductsRepository;

    @Inject
    ProductsListPresenter(ProductsRepository productsRepository) {
        mProductsRepository = productsRepository;
    }

    void getProductsList(Integer catId) {
        ifViewAttached(BaseView::startLoading);
        addDisposable(mProductsRepository
                .getProductsByCat(catId)
                .subscribe(
                        productsListResponse -> {
                            if(productsListResponse.isSuccessful() && productsListResponse.code()==200){
                                ifViewAttached(view -> view.onProductsListFetched(productsListResponse.body().getProductList()));
                            }
                            ifViewAttached(BaseView::stopLoading);

                        }, error->{
                            Timber.e(error);
                            ifViewAttached(BaseView::stopLoading);
                        }));
    }

    void getBestsellers() {
        ifViewAttached(BaseView::startLoading);
        addDisposable(mProductsRepository
                .getBestsellers()
                .subscribe(
                        productsListResponse -> {
                            if(productsListResponse.isSuccessful() && productsListResponse.code()==200){
                                ifViewAttached(view -> view.onProductsListFetched(productsListResponse.body().getProductList()));
                            }
                            ifViewAttached(BaseView::stopLoading);

                        }, error->{
                            Timber.e(error);
                            ifViewAttached(BaseView::stopLoading);
                        }));
    }
}

File: app/src/main/java/omniex/nl/omniex/ui/app/product/list/ProductsListPresenter.kt
package omniex.nl.omniex.ui.app.product.list

import omniex.nl.omniex.data.repository.ProductsRepository
import omniex.nl.omniex.ui.base.BasePresenter
import timber.log.Timber
import javax.inject.Inject

class ProductsListPresenter @Inject
internal constructor(private val mProductsRepository: ProductsRepository) : BasePresenter<ProductsListView>() {

    internal fun getProductsList(catId: Int?) {
        ifViewAttached{it.startLoading()}
        addDisposable(mProductsRepository
                .getProductsByCat(catId)
                .subscribe(
                        { productsListResponse ->
                            if (productsListResponse.isSuccessful() && productsListResponse.code() === 200) {
                                ifViewAttached { view -> view.onProductsListFetched(productsListResponse.body()!!.productList!!) }
                            }
                            ifViewAttached{it.stopLoading()}

                        }, { error ->
                    Timber.e(error)
                    ifViewAttached{it.stopLoading()}
                }))
    }

    internal fun getBestsellers() {
        ifViewAttached{it.startLoading()}
        addDisposable(mProductsRepository
                .bestsellers()
                .subscribe(
                        { productsListResponse ->
                            if (productsListResponse.isSuccessful() && productsListResponse.code() === 200) {
                                ifViewAttached { view -> view.onProductsListFetched(productsListResponse.body()!!.productList!!) }
                            }
                            ifViewAttached{it.stopLoading()}

                        }, { error ->
                    Timber.e(error)
                    ifViewAttached{it.stopLoading()}
                }))
    }
}
